### ISSUE TITLE
Make markup grammar LALR(1)

### ISSUE THREAD

**Issue by kemitchell** [commonform/commonform-new-markup#1]
@anseljh, I'd like to get your input on a possible change to the markup format. The idea is to make it easier to generate parsers, especially with Bison/Jison & friends.

The markup format isn't context-free, in that indentation is significant, but I've done some promising experiments emitting INDENT and DEDENT tokens from the lexical analyzer, [Python style](https://docs.python.org/2/reference/grammar.html). I read somewhere that Python is LL(2), and I think current Common Form markup is LL(2), as well, but I'd like to avoid hand-coding a recursive descent parser if I can.

One current feature that stumps single-token-lookahead parsers is child form headings:

```
\\ No heading. Just content.
# tokens: SLASHES TEXT

This is a Heading \\ Content, too!
# TEXT SLASHES TEXT

!! No heading. Just content.
# tokens: BANGS TEXT

This is a Heading !! Content, too!
# TEXT BANGS TEXT
```

Here's a proposal that I think does the job:

```
\\ Content
# SLASHES TEXT

__Heading__ \\ Content
# UNDERSCORES TEXT UNDERSCORES SLASHES TEXT

!! Conspicuous content
# BANGS TEXT

__Heading__ !! Conspicuous content
# UNDERSCORES TEXT UNDERSCORES BANGS TEXT
```

How does that strike you?

**Comment by kemitchell** [comment]
[Implemented with Jison](https://github.com/commonform/commonform-new-markup/blob/6f97e6ce798f0aa355f25f395e4c48b452da1d30/parser.jison)

**Comment by kemitchell** [comment]
@anseljh, I've implemented a second proposal with a bit less cruft:

```
\\\\ Content

\\!! Conspicuous content

\\ Heading  \\ Content

\\ Heading !! Conspicuous content
```

**Comment by tbrooke** [comment]
So is indentation remaining significant or is there another way to signify a child?

I have been looking at doing a parser just to be sure I have the grammar right

**Comment by kemitchell** [comment]
All the work here on commonform-new-markup is experimental. If and when it feels complete, I'll release it as new major version of commonform-markup.

I fully expect indentation will remain significant. The only indentation related question I'm still openly considering is whether text (not in children) should be indented to the same column as the start of the containing form or one level deeper.

**Comment by kemitchell** [comment]
@tbrooke sorry if I answered the wrong question.

The goal is an LALR(1) parser for Bison and friends. That means the grammar has to be context-free.

Line indentation isn't context-free. But we can cheat by doing all the context-sensitive processing in the lexer and emitting tokens (`INDENT` and `OUTDENT`) for the parser.

It's a trick I know best from Python's grammar, which IIRC is LL(2) and hand-coded, but uses the same approach. Conceptually, it's a bit like reading indentation ahead of time and replacing indents with `{` and outdents with `}`, and using those tokens, C-style, to delineate blocks.

Here is the most recent tokenizer code for emitting indentation tokens:

https://github.com/commonform/commonform-new-markup/blob/single-backslash/tokenize.js#L51

Here they are being used in the grammar around series of child forms:

https://github.com/commonform/commonform-new-markup/blob/single-backslash/parser.jison#L34

**Comment by anseljh** [comment]
Compilers are a bit beyond my comfort zone, but I took a look and things mostly made sense.

Am I correct that this would be valid new-style markup under #2?

``` text
# Comments are great.

So is introductory text, unless it includes silly words like "WHEREAS."

    \ Heading With No Child. \
    \ Heading With \ some child text.
    \ Heading With \
        \\ children
        \\ later,
        \\ including
        \!! conspicuous children!

Concluding text can be great, too.
```

**Comment by kemitchell** [comment]
Almost. Line 5 of your example contains a heading with no content, which can't exist in a valid form.

Slightly revised:

``` commonform
# Comments are great.

So is introductory text, unless it includes silly words like "WHEREAS."

    \ Heading With No Child. \ Must have content.
    \ Heading With \ some child text.
    \ Heading With \
        \\ children
        \\ later,
        \\ including
        \!! conspicuous children!

Concluding text can be great, too.
```

Parsed:

``` json
{ "content": [
    "So is introductory text, unless it includes silly words like \"WHEREAS.\"",
    { "heading": "Heading With No Child.",
      "form": {
        "content": [ " Must have content." ] } },
    { "heading": "Heading With",
      "form": {
        "content": [ " some child text." ] } },
    { "heading": "Heading With",
      "form": {
        "content": [
          { "form": {
              "content": [ " children" ] } },
          { "form": {
              "content": [ " later," ] } },
          { "form": {
              "content": [ " including" ] } },
          { "form": {
              "conspicuous": "yes",
              "content": [ " conspicuous children!" ] } } ] } },
    "Concluding text can be great, too." ] }
```

**Comment by anseljh** [comment]
:+1:

**Comment by kemitchell** [comment]
Closed by #2
